ref/c/segwit_addr.c

```diff
@@ -189,21 +189,44 @@ int segwit_addr_encode(char *output, const char *hrp, int witver, const uint8_t
     return bech32_encode(output, hrp, data, datalen, enc);
 }
 
-int segwit_addr_decode(int* witver, uint8_t* witdata, size_t* witdata_len, const char* hrp, const char* addr) {
+int segwit_addr_decode_detailed(int* witver, uint8_t* witdata, size_t* witdata_len,
+                char *hrp_out, bech32_encoding* enc_out, const char* addr)
+{
     uint8_t data[84];
     char hrp_actual[84];
     size_t data_len;
+
     bech32_encoding enc = bech32_decode(hrp_actual, data, &data_len, addr);
     if (enc == BECH32_ENCODING_NONE) return 0;
     if (data_len == 0 || data_len > 65) return 0;
-    if (strncmp(hrp, hrp_actual, 84) != 0) return 0;
     if (data[0] > 16) return 0;
-    if (data[0] == 0 && enc != BECH32_ENCODING_BECH32) return 0;
-    if (data[0] > 0 && enc != BECH32_ENCODING_BECH32M) return 0;
+
     *witdata_len = 0;
     if (!convert_bits(witdata, witdata_len, 8, data + 1, data_len - 1, 5, 0)) return 0;
     if (*witdata_len < 2 || *witdata_len > 40) return 0;
     if (data[0] == 0 && *witdata_len != 20 && *witdata_len != 32) return 0;
+
     *witver = data[0];
+
+    if(hrp_out) {
+        strncpy(hrp_out, hrp_actual, 20);
+    }
+    if(enc_out) *enc_out = enc;
```

Comment:
I don't like this. I cannot imagine where a caller is interested in valid or not, without caring if it's bech32 or bech32m.

Reply:
I can make the `enc_out` argument required